source/Halibut/Transport/Caching/ResponseCache.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.Caching;
+using Halibut.ServiceModel;
+using Halibut.Transport.Protocol;
+
+namespace Halibut.Transport.Caching
+{
+    /// <summary>
+    /// Allows the Response Caching behaviour to be modified for Error responses.
+    /// By default all errors responses are not cached.
+    /// </summary>
+    /// <param name="responseMessage">The error response message</param>
+    /// <returns>True to allow the error response message to be cached, otherwise false.</returns>
+    public delegate bool OverrideErrorResponseMessageCachingAction(ResponseMessage responseMessage);
+
+    internal class ResponseCache
+    {
+        readonly MemoryCache responseMessageCache = new("ResponseMessageCache");
+
+        public ResponseMessage GetCachedResponse(ServiceEndPoint endPoint, RequestMessage request, MethodInfo methodInfo)
+        {
+            var responseCanBeCached = CanBeCached(methodInfo);
+            if (!responseCanBeCached) return null;
+
+            var cacheKey = GetCacheKey(endPoint, methodInfo, request);
+            var wrapper = responseMessageCache.GetCacheItem(cacheKey)?.Value as CacheItemWrapper;
+
+            return wrapper?.ResponseMessage;
+
+        }
+
+        public void CacheResponse(ServiceEndPoint endPoint, RequestMessage request, MethodInfo methodInfo, ResponseMessage response, OverrideErrorResponseMessageCachingAction overrideErrorResponseMessageCachingAction)
+        {
+            var responseCanBeCached = CanBeCached(methodInfo, response, overrideErrorResponseMessageCachingAction);
+            if (!responseCanBeCached) return;
+
+            FixMissingHalibutRuntimeProcessIdentifier(response);
+
+            var cacheKey = GetCacheKey(endPoint, methodInfo, request);
+            var cacheDuration = GetCacheDuration(methodInfo);
+
+            var wrapper = new CacheItemWrapper
+            {
+                HalibutRuntimeUniqueIdentifier = response.HalibutRuntimeProcessIdentifier!.Value,
+                ResponseMessage = response,
+                EndPoint = endPoint
+            };
+            responseMessageCache.Add(cacheKey, wrapper, new CacheItemPolicy { AbsoluteExpiration = DateTimeOffset.UtcNow.AddSeconds(cacheDuration) });
+        }
+
+        public void InvalidateStaleCachedResponses(ServiceEndPoint endPoint, ResponseMessage response)
+        {
+            FixMissingHalibutRuntimeProcessIdentifier(response);
+
+            var cacheItems = responseMessageCache.ToList();
+
+            foreach (var item in cacheItems)
```

Comment:
This probably needs to be faster, since scanning 24k tentacle entries on each get capabilities call will be costly.

Computer can do 3000000000 cycles per second. Reduce that to 10ms latency so 30,000,000 cycles per 10ms.
Processing 24K tentacles so 1250 cycles to check each entry. If it takes more than that it will take more then 10ms on each call to invalidate the cache.

this PR improves on that significantly https://github.com/OctopusDeploy/Halibut/pull/242/files by only looking to invalidate entries within the service.